src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticQueue.cs

```diff
@@ -0,0 +1,211 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+
+namespace Microsoft.CodeAnalysis.Diagnostics
+{
+    internal abstract class DiagnosticQueue
+    {
+        public abstract bool TryComplete();
+        public abstract bool TryDequeue(out Diagnostic d);
+        public abstract void Enqueue(Diagnostic diagnostic);
+
+        // Methods specific to CategorizedDiagnosticQueue
+        public abstract void EnqueueLocal(Diagnostic diagnostic, DiagnosticAnalyzer analyzer, bool isSyntaxDiagnostic);
+        public abstract void EnqueueNonLocal(Diagnostic diagnostic, DiagnosticAnalyzer analyzer);
+        public abstract ImmutableArray<Diagnostic> GetLocalSyntaxDiagnostics(DiagnosticAnalyzer analyzer);
```

Comment:
It probably can do so, but I instead followed your other suggestion and renamed these methods to DequeueXXX to make the functionality more explicit.